### DISCUSSION TITLE
Beatmap listing and chat overlay focus loop when signed out

### DISCUSSION THREAD

**Discussion by Joehuu** [ppy/osu#16797] [Q&A] [answered]
![osu_2022-02-05_23-55-13](https://user-images.githubusercontent.com/35318437/152672286-5976d2b7-7934-4a3b-9b87-438a6b37a97c.jpg)

![osu_2022-02-06_00-00-48](https://user-images.githubusercontent.com/35318437/152672391-ffe01e80-56d2-45d7-86af-9376c4985c99.jpg)

**Comment by frenzibyte** [comment]
Appears to be going through the following cycle:
 1. `BeatmapListingOverlay`/`ChatOverlay` receives focus from becoming visible.
 2. `BeatmapListingOverlay`/`ChatOverlay` switches focus to their underlying `FocusedTextBox`.
 3. `InputManager` finds out `FocusedTextBox` is not visible in the scene ([via parent traversal in `unfocusIfNoLongerValid()`](https://github.com/ppy/osu-framework/blob/1a5b2b275d7061e41578b015db1ea39d78b41840/osu.Framework/Input/InputManager.cs#L909-L926)) and removes focus from it.
 4. `InputManager` looks for "top most requesting focus drawable" and ends up at `BeatmapListingOverlay`/`ChatOverlay`, and the cycle goes on.

A quick simple fix would be to avoid switching focus to the `FocusedTextBox` per usage, or improve `InputManager.ChangeFocus` to avoid switching focus if the drawable is not visible in the scene ([it appears to perform a simple presence check](https://github.com/ppy/osu-framework/blob/1a5b2b275d7061e41578b015db1ea39d78b41840/osu.Framework/Input/InputManager.cs#L382-L388), but doesn't do parent traversal unlike `unfocusIfNoLongerValid()`)

**Reply by bdach** [reply]
> A quick simple fix would be to avoid switching focus to the FocusedTextBox per usage

what does this mean even? does it mean that the "focused" text box *would no longer receive focus*?

the other option of not attempting to take focus if not present at least *sounds* sane.

**Reply by frenzibyte** [reply]
> > A quick simple fix would be to avoid switching focus to the FocusedTextBox per usage
> 
> what does this mean even? does it mean that the "focused" text box _would no longer receive focus_?

I meant to handle it locally per-usage by not sending focus to `FocusedTextBox` from the overlay if the user is not logged in. But yeah I'll go with the other option of not attempting to take focus.

**Comment by frenzibyte** [comment, marked as answer]
Will be resolved by https://github.com/ppy/osu-framework/pull/5017.
